packages/manifold/contracts/gachaclaims/ERC1155GachaLazyClaim.sol

```diff
@@ -0,0 +1,256 @@
+    return _getClaim(creatorContractAddress, instanceId);
+  }
+
+  /**
+   * See {IERC1155GachaLazyClaim-getClaimForToken}.
+   */
+  function getClaimForToken(
+    address creatorContractAddress,
+    uint256 tokenId
+  ) external view override returns (uint256 instanceId, Claim memory claim) {
+    instanceId = _tokenInstances[creatorContractAddress][tokenId];
+    claim = _getClaim(creatorContractAddress, instanceId);
+  }
+
+  function _getClaim(address creatorContractAddress, uint256 instanceId) private view returns (Claim storage claim) {
+    claim = _claims[creatorContractAddress][instanceId];
+    if (claim.storageProtocol == StorageProtocol.INVALID) revert IGachaLazyClaim.ClaimNotInitialized();
+  }
+
+  /**
+   * See {IGachaLazyClaim-mintReserve}.
+   */
+  function mintReserve(address creatorContractAddress, uint256 instanceId, uint32 mintCount) external payable override {
+    if (Address.isContract(msg.sender)) revert IGachaLazyClaim.CannotMintFromContract();
+    Claim storage claim = _getClaim(creatorContractAddress, instanceId);
+    // Checks for reserving
+    if (claim.startDate > block.timestamp) revert IGachaLazyClaim.ClaimInactive();
+    if (claim.endDate > 0 && claim.endDate <= block.timestamp) revert IGachaLazyClaim.ClaimInactive();
+    if (msg.value != (claim.cost + MINT_FEE) * mintCount) revert IGachaLazyClaim.InvalidPayment();
+    uint32 amountAvailable = claim.totalMax - claim.total;
+    if (amountAvailable == 0) revert IGachaLazyClaim.ClaimSoldOut();
+
+    // calculate the amount to reserve and update totals
+    uint32 amountToReserve = uint32(Math.min(mintCount, amountAvailable));
+    claim.total += amountToReserve;
+    _mintDetailsPerWallet[creatorContractAddress][instanceId][msg.sender].reservedCount += amountToReserve;
+
+    // Refund any overpayment
+    if (amountToReserve != mintCount) {
+      uint256 refundAmount = msg.value - (claim.cost + MINT_FEE) * amountToReserve;
+      _sendFunds(payable(msg.sender), refundAmount);
+    }
+    emit GachaClaimMintReserved(creatorContractAddress, instanceId, msg.sender, amountToReserve);
+  }
+
+  /**
+   * See {IGachaLazyClaim-deliverMints}.
+   */
+  function deliverMints(IGachaLazyClaim.Mint[] calldata mints) external override {
+    _validateSigner();
+    for (uint256 i; i < mints.length; ) {
+      Mint calldata mintData = mints[i];
+      Claim memory claim = _getClaim(mintData.creatorContractAddress, mintData.instanceId);
+      uint256[] memory tokenIds = new uint256[](1);
+      if (mintData.variationIndex > claim.tokenVariations || mintData.variationIndex < 1)
+        revert IGachaLazyClaim.InvalidVariationIndex();
+      tokenIds[0] = claim.startingTokenId + mintData.variationIndex - 1;
+      address[] memory receivers = new address[](mintData.recipients.length);
+      uint256[] memory amounts = new uint256[](mintData.recipients.length);
+      for (uint256 j; j < mintData.recipients.length; ) {
+        if (
```

Comment:
Reference accesses to maps are expensive.  If you are using/accessing more than once, you want to retrieve it into a storage variable.

MintDetails memory receiverMintDetails = _mintDetailsPerWallet[mintData.creatorContractAddress]

etc.

But this whole loop will likely change as per the below comments.